### ISSUE TITLE
Ginkgo 2.0 Proposal

### ISSUE THREAD

**Issue by onsi** [onsi/ginkgo#711]
The first commit to Ginkgo was on August 19th, 2013.  Much fruitful community feedback - and code cruft - have accumulated over the years.  Both can be addressed with a new major version of Ginkgo: Ginkgo 2.0.

The backlog capturing the work is here: https://www.pivotaltracker.com/n/projects/2493102

The original proposal for Ginkgo 2.0 has been pulled together as a google doc [here](https://docs.google.com/document/d/1h28ZknXRsTLPNNiOjdHIO-F2toCzq4xoZDXbfYaBdoQ/edit#). Note that the google doc is now stale - the Tracker backlog reflects the latest work most accurately.

V2 is in active development on the [ver2 branch](https://github.com/onsi/ginkgo/tree/ver2) and an uptodate changelog and migration doc is maintained on the branch [here](https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md)

If you use Ginkgo and are interested in what's in 2.0 please comment on this issue!

**Comment by onsi** [comment]
This proposal has been publicly available for three weeks.  After receiving some helpful feedback that has made it's way into the proposal I think it's time to start development.

**Comment by onsi** [comment]
I've put together a project to track progress toward v2:

https://github.com/onsi/ginkgo/projects/2

**Comment by onsi** [comment]
The first iteration of V2 is out on the [v2 branch](https://github.com/onsi/ginkgo/tree/v2).  There are a few new features on that branch, and most of the deprecations/removals lists in the V2 doc.  I'll be adding new functionality over the coming weeks/months.

This changelog + migration doc:
https://github.com/onsi/ginkgo/blob/v2/docs/MIGRATING_TO_V2.md

will be kept uptodate as I go.  I'll be removing the GitHub project in favor of a Pivotal Tracker backlog that I'm using (and will make public soon).

**Comment by mbalcomm** [comment]
I  just got this "warning" :
 You are using a custom reporter.  Support for custom reporters will likely be removed in V2.  Most users were using them to generate junit or teamcity reports and this functionality will be merged into the core reporter.  In addition, Ginkgo 2.0 will support emitting a JSON-formatted report that users can then manipulate to generate custom reports.

We have been using customer reporters for our kubernetes tests. Kubernetes test framework is based on Ginkgo. We have reports for storing test reports to databases that require specific data structures.  
 One reporter stores specific data into Prometheus. It would have to be a custom reporter. 
Another reporter does generate json format results files.  we could replace it with the proposed Json reporter, but it also lacks some functionalities we need.  For one thing is the ability to access detailed information, other than *SpecSummary
After reading the mitrating_to_v2, I am not very clear as how we can adopt new reporters and yet still be able to customize our specific data structures.  Any earlier V2 adopters out there? 
Thanks.

**Comment by onsi** [comment]
hey @mbalcomm - so glad to hear that that deprecation warning caught your eye.  I've been looking to get some early feedback :)

To confirm - you've read the [Reporting Infrastructure](https://github.com/onsi/ginkgo/blob/v2/docs/MIGRATING_TO_V2.md#improved-reporting-infrastructure) section in `MIGRATING_TO_V2`.

For your use case, assuming you don't need to emit test information in real time, I would expect that adding
```
var _ = ReportAfterSuite(func(report types.Report) {
	reporters.ReportViaDeprecatedReporter(prometheusReporter, report)
	reporters.ReportViaDeprecatedReporter(otherReporter, report)
})
```
would be a working starting point.

A few questions:

1. You mention "For one thing is the ability to access detailed information, other than *SpecSummary" - can you say more?  It's possible what's currently on V2 will work for your usecase.   It's possible you'll need a few more pieces that are currently planned for the V2 release which I'd be happy to describe to confirm they'll serve your usecase.
2. If it's publicly accessible, can you point me to the test suites and the custom reporters that you're referring to?
3. Would you be up for trying out V2 to give me some real-life feedback on this custom reporting piece?

Thanks!

**Comment by mbalcomm** [comment]
@onsi Thanks for your responses!
Yes, I have read the Migration_to_V2. and to answer your question: 
1.  The specific information I would like to capture in the report is "By". We use By() inside It for detailed test run information.
2. One reporter we are using is Allure Report(https://docs.qameta.io/allure/) since it does not have Golang plugin, we have to write our own. We like the Allure reporter's Step and attachment features. Allure Reporter is an open-source project. Although Kubernetes is open-source, currently our tests are not. 
3. I would be happy to try V2 out.

**Comment by onsi** [comment]
Hey @mbalcomm:

Thanks for your reply.  Some thoughts:

> 1.  The specific information I would like to capture in the report is "By". We use By() inside It for detailed test run information.
How are you currently capturing the information in `By`.  Currently, `By` simply writes to the GinkgoWriter but I'm thinking of evolving it to better integrate into the new reporting infrastructure which would make it easier to post-process programatically.

At the same time, the main reason that `By` exists is because it's hard for folks to write small granular `It` tests when their test cases are large and complex.  V2 will be introducing [ordered contexts](https://docs.google.com/document/d/1h28ZknXRsTLPNNiOjdHIO-F2toCzq4xoZDXbfYaBdoQ/edit#heading=h.lkc86l6lq1be) to make it possible to break out large complex `It`s into a collection of smaller `It`s that are guaranteed to run in order.

In any event, the behavior of `By` hasn't changed yet and when it does it will become better integrated into the new reporting infrastructure.

> 2. One reporter we are using is Allure Report(https://docs.qameta.io/allure/) since it does not have Golang plugin, we have to write our own. We like the Allure reporter's Step and attachment features. Allure Reporter is an open-source project. Although Kubernetes is open-source, currently our tests are not. 

Have you written a custom reporter to emit to Allure?  If so that should continue to work via `reporters.ReportViaDeprecatedReporter`.  If there's enough community interest I could also imagine making Allure's native XML format a first-class report target for Ginkgo so you would simply run `ginkgo -allure-report=out.xml` but that's not something I'm ready to prioritize yet.

The other piece of reporting infrastructure that is coming in V2 will be to support things like attachments.  So users can do stuff like:

```
It("...", func() {
    ....
    Report("key", value)
})
```
Where value can be any object.  Ginkgo will include the passed in key:value pair in the generated report allowing it to be exported as desired.

> 3. I would be happy to try V2 out. 
Great!  It's all on the v2 branch.

**Comment by jakeczyz** [comment]
Quick Comment: 

We are using a custom reporter and not ready to migrate to v2 until v2 is out of beta/stable.  So, in the mean time, whenever we run tests we get those several lines of deprecation warning text eating up screen space (and attention).  

It would be super helpful if there was a way to disable the deprecation warnings via flag or env var or otherwise. :)

**Comment by onsi** [comment]
hey @jakeczyz PTAL at #802 - you can now set an env var to silence deprecations

**Comment by bobdonat** [comment]
I am experimenting with v2 and I am interested in the test label feature.  It appears that this feature is not there yet.  Is this feature still planned for v2 and when will that feature be available?

**Comment by onsi** [comment]
Yep - still planned.  Summer schedules have slowed down development a bit so i'm having a harder time reasoning about the timeline but i'm hoping to get to that feature in the next couple 2ish months.

**Comment by chrisdoherty4** [comment]
We were using custom reporters to send information to TestRail. It looks like the `ReportAfterSuite()` capability will satisfy our needs though - thought I'd give the feedback none-the-less.

**Comment by onsi** [comment]
thanks @chrisdoherty4 !

**Comment by TaylorOno** [comment]
After the deprecation custom reports will the any built in reports allow for integration with Goland which seems to only play nice with the go test format?

Reports like https://github.com/SemanticallyNull/golandreporter which provide real time feedback

**Comment by onsi** [comment]
It would be fairly straightforward to use and/or port the golandreporter you reference.  In V2 ginkgo will support emitting a JSON-formatted report that other tooling can build on top off (whether on the goland/ide side or on the ginkgo side or some sort of shim in-between).

**Comment by TaylorOno** [comment]
I am just not clear how you would accomplish that without a way to emit go test formatted output in real time. Goland play test function will naively run these commands sending the output to test2json.
```
/home/taylor/go/go1.16.5/bin/go test -c -o /tmp/___go_test_github_com_TaylorOno_sample.test github.com/TaylorOno/sample #gosetup
/home/taylor/go/go1.16.5/bin/go tool test2json -t /tmp/___go_test_github_com_TaylorOno_sample.test -test.v -test.paniconexit0
```

**Comment by onsi** [comment]
Ah, I see.  For this I would imagine a library that hooks in via `ReportAfterEach`.  A simple `_` import would be enough to add the hook and goland can take it from there.

**Comment by rmohr** [comment]
We use in kubevirt custom reports to collect the current state of a cluster when a test fails. Just filed https://github.com/onsi/ginkgo/issues/823 when I realized that the reporters are going to be removed. I think `ReportAfterEach` can give us what we have right now. Did you also consider a `ReportBeforeAfterEach`, so that one can collect information without destroying the state which led to a failing test?

**Comment by onsi** [comment]
Like I said in #823 you can use `JustAfterEach` and have the same effect.  The same report object passed to `ReportAfterEach` is available via `CurrentSpecReport()` which can be called at any point.

**Comment by TaylorOno** [comment]
> Ah, I see. For this I would imagine a library that hooks in via `ReportAfterEach`. A simple `_` import would be enough to add the hook and goland can take it from there.

one more question about the golandreporter.  Previously `SpecWillRun` was being used, what would be the equivalent? or will we only be able to report on that after the spec has completed via the `ReportAfterEach`?

**Comment by onsi** [comment]
@TaylorOno yes, `SpecWillRun` is going away - there just weren't that many use-cases for it.  It can be replaced with `BeforeEach` with the only downside being that skipped specs will not run their `BeforeEach` (but will run their `ReportAfterEach`).  I can and may add a `ReportBeforeEach` after 2.0 ships.

**Comment by TaylorOno** [comment]
a `ReportBeforeEach` would nice since it would allow you to have Goland display a nice spinner for long running tests.  My real dream is having a ginkgo goland plugin.

**Comment by onsi** [comment]
Will add `ReportBeforeEach` to the backlog.  Agreed that the long-term answer here is a Ginkgo goland plugin!

**Comment by mat007** [comment]
Hi, is there a way to update to V2 with a Go modules `go get`?
The [Go documentation](https://go.dev/blog/v2-go-modules) suggest to have V2 in a sub-folder along side V1, but here it’s in a branch.

Sorry if this is a silly question and I’m missing something obvious 😬 
I looked at the migration guide but couldn’t find any advice there.

**Comment by blgm** [comment]
@mat007 it doesn't work with `go get` yet because the V2 branch hasn't had the import path modified here: https://github.com/onsi/ginkgo/blob/f772b55b9575950075d19f196e0b0f41fc893782/go.mod#L1

This is because the V2 branch is still under development. However if you want to try it out, you can:
```
git clone git@github.com:onsi/ginkgo.git
cd ginkgo
git switch v2
go install
```
This will put the V2 `ginkgo` command in `$GOBIN` or `$GOPATH/bin` which is typically in the `$PATH`.

We would welcome any feedback that you have. The installation process will definitely be improved!

**Comment by onsi** [comment]
Thanks @blgm - at this point I'm ready to call V2 beta.  I've updated the [V2 Migration Guide](https://github.com/onsi/ginkgo/blob/v2/docs/MIGRATING_TO_V2.md#using-the-beta) with instructions on how to download and use the beta as well as a (long!) list of improvements and changes that have landed.  Would love to get feedback!

**Comment by mat007** [comment]
> Update your go module to point to the local, V2, copy of Ginkgo:
>    cd /PATH/TO/YOUR/PACKAGE
>    go mod edit -replace github.com/onsi/ginkgo=/PATH/TO/GINKGO/V2

This is probably fine for a quick test run (pun intended 😁), but I was myself planning on switching a rather large project over to ginkgo v2 for all my coworkers and our CI.
I think I will probably fork ginkgo, move the v2 branch to a v2 folder, and update our project to point to my fork. Or is there a simpler solution?

**Comment by onsi** [comment]
Hey @mat007 and @blgm good news (I think!).  Go's module toolchain does support checking out branches however the branch name I'd picked (`v2`) was clashing with go mod's server support.  I've renamed the branch to `ver2`  and this has substantially simplified how you get access to version 2.  Just:

```bash
go get github.com/onsi/ginkgo@ver2
go install github.com/onsi/ginkgo/ginkgo
```

Doing this will check out a pseudo-version tied to the tip of the `ver2` branch.  Just rerun the commands to bump to the latest.  For the beta you won't need to update your import paths but you will need to once V2 goes GA and merges onto master.  @mat007 you should be able to switch over your project in this way - no need for v2 folders to play with the beta.

@bobdonat you'd asked about labels - they just landed on the `ver2` branch and I have preliminary documentation [here](https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#new-spec-labels).

Docs for using the beta in a project have been updated [here](https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#using-the-beta).

**Comment by KevinJCross** [comment]
Hi Im considering trying to add a feature where a file log of the time of each stage is added. Currently we have quite a few issues with slowness in our meduim sized test suites and to get some metrics as to which bits are taking the most time would be greatly advantagous.
I.e. 
CompileTimes
BeforeSuite
AfterSuite
BeforeEach's 
AfterEach's
JustBefore
and Its
Practically anywhere where time is spent inside or outside the test suite.
I recognise that most/all of it would be in the tests but to pinpoint which justbefore each or BeforeEach step is call ethe most and uses the most amount of time would help gather the information to optimise our test suites.

I would thikning along the lines of a csv file output styles with 
something like 
Description, average call time (s), total time (s), number of calls 
"Description"."context"."context".BeforeEach, 0.200,  20,  200

I was thinking I might be able to farm it out the profiles but it would be nice to have a summary option off the bat.

**Comment by bobdonat** [comment]
Will the JUnit generated report identify which tests are flaky along with the number of attempts

**Comment by onsi** [comment]
@bobdonat the JUnit schema (which... is a bit hard to find an official version of, though [this](https://llg.cubic.org/docs/junit/) is a good resource) doesn't have much of a mechanism for specifying per-spec information in a structured way.  I end up munging stuff into the existing fields.  So, currently, the JUnit report does not include # of attempts.

I'd love to encourage folks who are running large suites that need analytics like this to look at the new [JSON format](https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#improved-reporting-infrastructure) which has all the [spec information](https://github.com/onsi/ginkgo/blob/ver2/types/types.go#L107-L165) captured in a structured way.

@KevinJCross I like the idea and the problem you're trying to solve makes a lot of sense.  I'm going to open a separate issue so we can dig into some potential solutions.  I'd like to leverage the new primitives in 2.0 - I think we can build on top of the JSON format and [ReportEntries](https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#new-report-entries) to annotate specs with additional information like this.

**Comment by matt-royal** [comment]
So glad you're doing this Onsi! My team just upgraded for the `Serial` support, which is very useful.

I do have a help and/or feature request. Is it currently possible to add a `Label(...)` at the level of the entire test package? Right now I'm having to add the same annotation to every top-level `Describe` in the package instead. Thanks!

**Comment by onsi** [comment]
ooh that's a great point.  I can make it so you can add `Label` to `RunSpecs` - will work on it tomorrow :)

**Comment by marianopinto** [comment]
Hi @onsi, thanks a lot for the amazing work you are doing for the community. I just upgraded Ginkgo and integrated it in Jenkins. While I run the tests locally everything seems to be going pretty smooth with the Junit reports getting generated. In the CI I am seeing issues, in particular I am seeing this:
```[ReportAfterSuite] Autogenerated ReportAfterSuite for --junit-report
autogenerated by Ginkgo

  Begin Captured GinkgoWriter Output >>
    [ReportAfterSuite] TOP-LEVEL
      autogenerated by Ginkgo
  << End Captured GinkgoWriter Output

  Failed to generate JSON report:
  open junit.xml: read-only file system
  In [ReportAfterSuite] at: /Users/itmp2/go/pkg/mod/github.com/onsi/ginkgo@v1.16.6-0.20211115182015-aec53805973a/reporting_dsl.go:127```
  The file is meant to be dropped in the tmp folder of a lambda function with the Jenkins Junit plugin to then process it. I am sure I am missing something here, just wondering if you could please point me in the right direction? Ta

**Comment by onsi** [comment]
hey @marianopinto do you mind opening a separate issue so we can dig into the details?  can you share how you're running the tests (i.e. what are the flags you're passing in to the `ginkgo` cli?)

**Comment by onsi** [comment]
hey @matt-royal you can now annotate suites with labels by passing the `Label()` decorator into `RunSpecs()`.  Give it a try and let me know if it suits y'all's need!

**Comment by onsi** [comment]
Heads up that 2.0.0-rc2 is out now.  I've formally bumped the major version in go.mod so you'll need to change your import path to `github.com/onsi/ginkgo/v2`.

I've also tagged the release which should make it a bit easier to pin to the rc.  I'll be tagging subsequent RCs if there are any.

Updated instructions for fetching and using the RC are [here](https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#using-the-beta).

**Comment by LittleFox94** [comment]
Hi,
I'm having an issue with an `Ordered` `Context` and not sure if this is intended or worthy for it's own ticket:

My test (which is an e2e/integration test) has roughly this structure:
```
Describe
|-- BeforeEach(create object of another API we need for our tests)
|-- AfterEach(delete object created in BeforeEach) // problematic AfterEach 
|-- Context("working on a fresh object", Ordered)
| |-- It("successfully creates the object")
| |-- It("successfully deletes the object")
|-- Context("working with an existing test Object")
| |-- BeforeEach("create test Object")
| |-- AfterEach("delete test Object")
| |-- It("finds Object when listing")
| |-- It("retrieves Object with expected values")
...
```

The problem is the problematic `AfterEach` being run after every `It` in the `Ordered` `Context`, I expected it to run once for the `Ordered` container. The semantic for `AfterEach` is to run for every `It`, but this might not be the best when having a nested Ordered container? Is there a way to make this work with the current API? Maybe having a `AfterEachDirectChild` would be useful, running something for each `It` or container directly in the current one?

**Comment by LittleFox94** [comment]
first: sorry for not having answered to your first reply, that task was pushed aside for now and the notification email still in inbox to process at some point ^^'

In my case `BeforeEach` and `AfterEach` each running once for the `Ordered` container would work great and I think they should always work as pair with same behavior. I'm not sure though if changing the semantics of `Before/AfterEach` is good or if adding new hooks or optional parameters might be better - feels like a kinda huge semantic change to me

**Comment by onsi** [comment]
Thanks.  I don't think it's a super huge change since `Ordered` containers are a new construct and really behave like large Its composed of more granular subunits.

**Comment by LittleFox94** [comment]
Hm yes, that makes sense

**Comment by kgadams** [comment]
Hello, I am looking for a way to integrate some reporting about the progress of Ginkgo into our own log file.
Right now we get ginkgo messages on stdout (including the configurable error reporting) and logging from our own coding in a dedicated logfile.
Now I would like to add logfile entries about all the actions that ginkgo takes, so that I can match this more easily.
I had started to implement a customer reporter for this, as the hooks in the interface look promising. Now I find that this is being deprecated.
Is there another mechanism in Ginkgo V2 that will enable this functionality? The migration guide did not give me any hints.

**Comment by onsi** [comment]
hey @kgadams - the topic of V2's reporting infrastructure is covered in the migration guide [here](https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#improved-reporting-infrastructure).

In brief: you can probably accomplish what you want using [`ReportBeforeEach` and `ReportAfterEach`](https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#capturing-report-information-about-each-spec-as-the-test-suite-runs).  However if you intend to run your specs in parallel and want to emit logging information into a single global log-file you will run into issues with interleaved output.  There are several strategies to manage this that I'd be happy to get into - but we should probably open a different issue for that.

Lastly - the new V2 documentation goes into the reporting infrastructure in a bit more depth than the migrating guide.  You can check that out [here](https://github.com/onsi/ginkgo/blob/ver2/docs/index.md#reporting-infrastructure).

**Comment by kgadams** [comment]
> In brief: you can probably accomplish what you want using [`ReportBeforeEach` and `ReportAfterEach`]

Ok, thank you, that looks good to me.

> Lastly - the new V2 documentation goes into the reporting infrastructure in a bit more depth than the migrating guide. You can check that out [here](https://github.com/onsi/ginkgo/blob/ver2/docs/index.md#reporting-infrastructure).

Thanks again 👍

**Comment by onsi** [comment]
hey @LittleFox94 I changed tack and ended up leaving the default seating of `*Each` untouched.  You can, however, now decorate `BeforeEach` and `AfterEach` with the new `OncePerOrdered` decorator.  Please bump to the latest and give it a try!

**Comment by onsi** [comment]
Hey all - we're super close to the GA here.  I just pushed out one last release candidate.  rc3 is now out and includes the new [`OncePerOrdered` decorator](https://github.com/onsi/ginkgo/blob/ver2/docs/index.md#setup-around-ordered-containers-the-onceperordered-decorator).  If you've been using the 2.0 RCs **THANK YOU**!  The feedback has been invaluable!  Please give this latest release candidate a try (especially if you're using `Ordered` containers).

I hope to have the 2.0 GA out before the new year!

**Comment by LittleFox94** [comment]
> hey @LittleFox94 I changed tack and ended up leaving the default seating of `*Each` untouched. You can, however, now decorate `BeforeEach` and `AfterEach` with the new `OncePerOrdered` decorator. Please bump to the latest and give it a try!

looks good for me, I like this solution :)

**Comment by onsi** [comment]
🎉🎉 🎉 

After many months - and just before we all ring in the new year -  I'm super happy to share that Ginkgo 2.0 is now officially GA.  The v2.0.0 code is now tagged and on master.  Huge thanks to the many users who gave the betas and RCs a try and provided so much valuable feedback to make this release what it is today.

🎉🎉 🎉
